### DISCUSSION TITLE
课后作业: 完成的同学把 github 链接发到下面

### DISCUSSION THREAD

**Discussion by luckyyang** [coset-io/zkp-academy#20] [General]
一共有三个小作业, 目标是补充 [baby plonk](https://github.com/Antalpha-Labs/baby-plonk) 中缺失的代码, 包含 setup, prover 和 verifier 相关的部分.

作业目标:
- 作业 1: 完成 setup.py 的 SRS 生成
- 作业 2: prover 完成多项式约束的构建
- 作业 3: verifier 完成验证
- 附加作业: 添加 plonkup

可以结合上课的讲义和代码来完成.

baby plonk 代码: https://github.com/Antalpha-Labs/baby-plonk

### 作业 1: 完成 setup.py 的 SRS 生成
1. 切换到 `homework_1` 分支
2. 打开 `setup.py`
3. 完成 `TODO` 部分代码
4. 确保运行测试通过: `poetry run python test.py`

测试运行通过的 log:
```shell
$ poetry run python test.py

Finished to generate structured reference string
Permutation accumulator polynomial successfully generated
Generated the quotient polynomial
Generated final quotient witness polynomials
Prover test success
Beginning verifier test
Done KZG10 commitment check for a_eval polynomial
Done KZG10 commitment check for b_eval polynomial
Done KZG10 commitment check for c_eval polynomial
Done KZG10 commitment check for z_eval polynomial
Done KZG10 commitment check for zw_eval polynomial
Done KZG10 commitment check for t_eval polynomial
Done KZG10 commitment check for ql_eval polynomial
Done KZG10 commitment check for qr_eval polynomial
Done KZG10 commitment check for qm_eval polynomial
Done KZG10 commitment check for qo_eval polynomial
Done KZG10 commitment check for qc_eval polynomial
Done KZG10 commitment check for s1_eval polynomial
Done KZG10 commitment check for s2_eval polynomial
Done KZG10 commitment check for s3_eval polynomial
Done equation check for all constraints
Verifier test success
```

### 作业 2: prover 完成多项式约束的构建
1. 切换到 `homework_2` 分支
2. 打开 `prover.py`
3. 完成 `TODO` 部分代码
4. 确保运行测试通过: `poetry run python test.py`

### 作业 3: verifier 完成验证
1. 切换到 `homework_3` 分支
2. 打开 `verifier.py`
3. 完成 `TODO` 部分代码
4. 确保运行测试通过: `poetry run python test.py`

### 附加作业: 添加 plonkup
按照课程内容和讲义, 给 baby plonk 添加 plonkup

**Comment by doutv** [comment]
已完成作业123 https://github.com/doutv/baby-plonk

**Comment by keroro520** [comment]
已完成作业123

- homework_1_solution: https://github.com/keroro520/baby-plonk/pull/1/files
- homework_2_solution: https://github.com/keroro520/baby-plonk/pull/2/files
- homework_3_solution: https://github.com/keroro520/baby-plonk/pull/3/files

**Reply by mjzk** [reply]
As you said "the results of Z(wX) g(X) - Z(X) f(X) and z(X) f(X) - z(wX) g(X) are different", but the constraint said Z(wX) g(X) - Z(X) f(X)=0, and then z(X) f(X) - z(wX) g(X)=0. So if the constraint holds always, should they be same? and why should you use one rather than use the other one?

**Reply by keroro520** [reply]
z(X) f(X) - z(wX) g(X) = 0 only when X in $\mathbb{H}$

When we expand the scope of X to $\mathbb{F}$, the equation is not always right, so that we can use schwartz-zippel

**Reply by keroro520** [reply]
![IMG_4924](https://github.com/Antalpha-Labs/zkp-academy/assets/1870648/bfc23819-effb-4cec-bd96-c8c17b6b3736)

**Comment by liqi16** [comment]
Simba已完成作业123: https://github.com/liqi16/baby-plonk

**Reply by mjzk** [reply]
seemly no change?

**Reply by liqi16** [reply]
I added my code to the TODO lines in the homework_1,2,3 branch. Could you possibly see the change?

**Comment by mjzk** [comment]
MJ's homework 1/2/3: https://github.com/mjzk/baby-plonk-homework

**Comment by jonathanxuu** [comment]
jonathan's homework 1/2/3: https://github.com/jonathanxuu/baby-plonk

**Comment by sgz13140** [comment]
sgz's homework1/2/3: https://github.com/sgz13140/baby-plonk

**Comment by 0xxEric** [comment]
Eric's homework1/2/3: 
https://github.com/0xxEric/baby-plonk

**Comment by sunjianping-alt** [comment]
已完成作业https://github.com/sunjianping-alt/baby-plonk

**Comment by Wpippen** [comment]
pippen's homework 1/2/3
https://github.com/PipengWang/baby-plonk

**Comment by ddl-hust** [comment]
ddl's  homework:
https://github.com/ddl-hust/baby-plonk

**Comment by tmors** [comment]
Dust's homework:
https://github.com/tmors/baby-plonk

**Comment by meloalright** [comment]
MeloAlright's homework

Just tried my best to get hands dirty and understand other classmates' commit to make it

[https://github.com/meloalright/baby-plonk](https://github.com/meloalright/baby-plonk)

**Comment by Synia-L** [comment]
Synia's homework:
https://github.com/Synia-L/baby-plonk-hk

**Comment by hattiepwvi** [comment]
Eta‘ s homework: Verifier test success
https://github.com/ETAAcademy/ETAAcademy-ZK-Dev/tree/main/02_baby_plonk

**Reply by hattiepwvi** [reply]
Eta‘ s homework 1/2/3: 
https://github.com/hattiepwvi/baby-plonk
Try to add lookup table ing:
https://github.com/ETAAcademy/ETAAcademy-ZK-Dev/tree/main/02_baby_plonk

**Comment by max65536** [comment]
max's homework: 
https://github.com/max65536/baby-plonk/tree/homework_3

**Reply by max65536** [reply]
https://github.com/max65536/baby-plonk

**Comment by yct21** [comment]
已完成作业：

- [homework 1](https://github.com/yct21/baby-plonk/commit/0081ae78f2c61e495117b250275e49f278ae009f)
- [homework 2](https://github.com/yct21/baby-plonk/commit/b886d7dda4e5429f82c11716bdc889f4fb1eae6d)
- [homework 3](https://github.com/yct21/baby-plonk/commit/898594eeae96b02dbc0a639a9c6aac1ffb6a72af)

**Comment by cryptosupu** [comment]
https://github.com/cryptosupu/baby-plonk

**Comment by baidang201** [comment]
Li Smith‘ s homework https://github.com/baidang201/baby-plonk-work.git

**Comment by max65536** [comment]
max's homework: https://github.com/max65536/baby-plonk/tree/plonkup
尝试写了下附加作业plonkup，目前只写了setup, round1, round2

**Comment by kkdomono** [comment]
林's homework:
homework1: https://github.com/kkdomono/baby-plonk/pull/1
homework2: https://github.com/kkdomono/baby-plonk/pull/2
homework3: https://github.com/kkdomono/baby-plonk/pull/3

**Comment by 0xxEric** [comment]
补充了附加作业plonkup，实现了查询向量的一个Demo，详见：
https://github.com/0xxEric/baby-plookup

**Comment by Wenzhou-Yang** [comment]
ywz's homework: https://github.com/Wenzhou-Yang/baby-plonk

**Comment by hellojason3** [comment]
ddl   homework  https://github.com/hellojason3/baby-plonk

**Comment by viaNull** [comment]
完成作业： https://github.com/viaNull/baby-plonk

**Comment by bofu2007** [comment]
bofu zhang (bogu)完成作业：https://github.com/bofu2007/baby-plonk

**Comment by whgreate** [comment]
完成作业123：https://github.com/whgreate/baby-plonk

**Comment by MADAOwww** [comment]
游云完成作业： https://github.com/MADAOwww/baby-plonk

**Comment by hawkinggg** [comment]
完成作业123  https://github.com/hawkinggg/baby-plonk/

**Comment by leowei1234567** [comment]
完成作业： https://github.com/leowei1234567/baby-plonk

**Comment by JustXxx** [comment]
Bill    https://github.com/JustXxx/baby-plonk

**Comment by MartinYeung5** [comment]
完成作业: https://github.com/MartinYeung5/baby-plonk

**Comment by code-fool** [comment]
完成作业 ： https://github.com/code-fool/baby-plonk.git

**Comment by bufrr** [comment]
完成作业：https://github.com/bufrr/baby-plonk/
